math/src/main/java/smile/math/matrix/SparseMatrix.java

```diff
@@ -175,11 +180,204 @@ public int size() {
     public double[] values() {
         return x;
     }
-    
+
+    /**
+     * Returns the row index
+     */
+    public int[] getRowIndex() {
+        return rowIndex;
+    }
+
+    /**
+     * Returns the column index
+     */
+    @SuppressWarnings("WeakerAccess")
+    public int[] getColIndex() {
+        return colIndex;
+    }
+
+    /**
+     * Calls a lambda with each non-zero value. This will be a bit faster than using a stream
+     * because we can avoid boxing up the coordinates of the element being processed, but it
+     * will be considerably less general.
+     * <p>
+     * Note that the action could be called on values that are either effectively or actually
+     * zero. The only guarantee is that no values that are known to be zero based on the
+     * structure of the matrix will be processed.
+     *
+     * @param action Action to perform on each non-zero, typically this is a lambda.
+     */
+    @SuppressWarnings("WeakerAccess")
+    public void foreachNonzero(MatrixElementConsumer action) {
```

Comment:
How about just `foreach` as the method name?

Reply:
Again, this would be misleading. It doesn't iterate over all elements of the matrix. It only iterates over non-zeros.

Reply:
Understood. We prefer short method name and the documents state clearly that it iterates only nonzero values. We don't need make code as verbose as Objective-C.

Reply:
This isn't verbose. It is self-documenting.

Reply:
In application code, the javadoc won't carry across, but the name will. Having the Nonzero as part of the name is important for readers of the code to understand the function.